.metadata/generate.proj

```diff
@@ -0,0 +1,37 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project Sdk="Microsoft.Windows.WinmdGenerator/0.38.19-preview">
+    <PropertyGroup Label="Globals">
+        <OutputWinmd>../.windows/winmd/Windows.Win32.Graphics.Direct3D12.winmd</OutputWinmd>
+        <WinmdVersion>1.608.0</WinmdVersion>
+        <IdlDir>../include/directx</IdlDir>
+    </PropertyGroup>
+    <ItemGroup>
+        <Idls Include="$(IdlDir)/dxgicommon.idl;$(IdlDir)/dxgiformat.idl;$(IdlDir)/d3dcommon.idl;$(IdlDir)/d3d12.idl" />
+        <!-- ;$(IdlDir)/d3d12sdklayers.idl -->
+
+        <DxgiHeaders Include="$(CompiledHeadersDir)\dxgiformat.h;$(CompiledHeadersDir)\dxgicommon.h" />
+        <Direct3D12Headers Include="$(CompiledHeadersDir)\d3d12.h" />
+
+        <Partition Include="dxgicommon.cpp">
+            <TraverseFiles>@(DxgiHeaders)</TraverseFiles>
+            <ExcludeFromCrossarch>true</ExcludeFromCrossarch>
+            <Namespace>Windows.Win32.Graphics.Dxgi.Common</Namespace>
+        </Partition>
+
+        <Partition Include="d3dcommon.cpp">
+            <TraverseFiles>$(CompiledHeadersDir)\d3dcommon.h</TraverseFiles>
+            <ExcludeFromCrossarch>true</ExcludeFromCrossarch>
+            <Namespace>Windows.Win32.Graphics.Direct3D</Namespace>
+        </Partition>
+
+        <Partition Include="d3d12.cpp">
+            <!-- TODO: DXGI headers are here to allow d3d12 to find DXGI_FORMAT,
+                 but that does cause the type to _also_ show up under the Direct3D12
+                 namespace.
+            -->
+            <TraverseFiles>@(DxgiHeaders);@(Direct3D12Headers)</TraverseFiles>
+            <ExcludeFromCrossarch>true</ExcludeFromCrossarch>
+            <Namespace>Windows.Win32.Graphics.Direct3D12</Namespace>
```

Comment:
@mikebattista any idea how I can make the `Direct3D12` namespace reference things from the `Dxgi` namespace in the parent `Graphics` namespace, and make sure this keeps compiling?

Reply:
Anything included in `TraverseFiles` will be included in the metadata. If you just need to reference DXGI to compile but don't want to include those headers in the Direct3D12 namespace, you typically would add a reference in `main.cpp` but leave them out of `TraverseFiles`. See https://github.com/microsoft/win32metadata/blob/9f7f4fa28e46c85e04ca0f11f89f1196dd07ce8d/sources/GeneratorSdk/samples/DWriteCoreSample/DWriteCoreSample.proj#L15-L21 for an example.

Reply:
@mikebattista thanks, that's roughly what I expected but didn't add `#include <dxgiformat.h>` to `main.cpp` since it already resides in `d3d12.h` - or is the parser doing something special and only caring about top-level includes, since "everything else" seems to be fine?

Reply:
@mikebattista I just did that in https://github.com/MarijnS95/DirectX-Headers/commit/490dcbec2f5d61075c321db11ad0a6e47ff4e5f3, but it still fails in the exact same way: https://github.com/MarijnS95/DirectX-Headers/actions/runs/3517519822/jobs/5895385933

Reply:
Can you check the generated cs file? Is it missing the using statement you would need to bring in that enum?

Reply:
There is a `using Windows.Win32.Graphics.Dxgi;`, but not `.Common` which I think is where this type should reside.

Maybe the problem is caused by having to pull in "stock" win32 .winmd which _also_ defines (part of) these APIs?

Reply:
https://github.com/MarijnS95/DirectX-Headers/actions/runs/3518395377 @mikebattista this artifact contains all `*.cs` files, see `.metadata/obj/generated/common/Windows.Win32.Graphics.Direct3D12.cs` where it `uses` `Dxgi` but not `Dxgi.Common` where `DXGI_FORMAT` is defined (`.metadata/obj/generated/common/Windows.Win32.Graphics.Dxgi.Common.cs`).

Reply:
So in the win32metadata repo there is https://github.com/microsoft/win32metadata/blob/9f7f4fa28e46c85e04ca0f11f89f1196dd07ce8d/generation/WinSDK/scraper.header.txt#L50 which lists the using statements that will be included in all generated .cs files. As you can see, Dxgi.Common is not currently listed there. We could add it there and the tooling would pick it up in the next release, but I'm curious if you can override this in the meantime.

@riverar have you encountered this type of error before? Is there a way to include additional `scraper.header.txt` includes that the tooling doesn't natively support?

Reply:
Quite curious that `Windows.Win32.winmd` from the NuGet package _does have_ that `using` statement. Is this perhaps related?

https://github.com/microsoft/win32metadata/blob/9f7f4fa28e46c85e04ca0f11f89f1196dd07ce8d/generation/WinSDK/requiredNamespacesForNames.rsp#L153

Reply:
That's not a using statement. That's just assigning the type to that namespace. For cs files in a different namespace, you'll still need a using statement to be able to resolve types from another namespace.

Reply:
@mikebattista `--requiredNamespaceForName` reads to me as "if you encounter this type/name/constant, you need to include this namespace via `using`". After all not even the manual `.cs` file includes it:

https://github.com/microsoft/win32metadata/blob/9f7f4fa28e46c85e04ca0f11f89f1196dd07ce8d/generation/WinSDK/manual/Direct3D12.cs

I'm glad that I tried in https://github.com/MarijnS95/DirectX-Headers/commit/eb22710, since this actually fixes it! The `using` statement is included in the `.modified.cs` file.

Reply:
Interesting. I'm not sure yet if your fix is a happy coincidence or by design behavior, but glad you're unblocked.

requiredNamespaceForName is intended for scenarios where (1) a header is assigned to a namespace, but you want to redirect some types within to another namespace or (2) a type is defined in two different headers that would be assigned to different namespaces, and you manually assign the type to the desired namespace to ensure determinism.

Reply:
Sure, let me know in case this is just a coincidence, and is going to be cleaned up upstream since it seems to be relying on this behaviour now, too (instead of adding to `scraper.header.txt` as you suggested).

Reply:
I added Dxgi.Common with https://github.com/microsoft/win32metadata/commit/273d404e3e851e31766a884c1bb9a14d3606570f. That namespace exists in the SDK metadata as well so there's no harm in adding it.

Reply:
Perfect, thanks! I'll unresolve this conversation to not forget about removing the workaround.

Seems the `# region dxgi.h` may indicate what you mentioned above: that this type was pulled out from `Dxgi` into `Dxgi.Common`? After all the heaader is split into `dxgicommon.h` and `dxhiformat.h` (and IDL) in this repo.

Reply:
> @riverar have you encountered this type of error before? Is there a way to include additional `scraper.header.txt` includes that the tooling doesn't natively support?

The tooling will pick up `scraper.header.txt` in the msbuild project directory and consolidate it with the others (e.g. that ship with the SDK tooling, the base header, etc.)

Reply:
@riverar seems like that file in the project directory is default-included here?:

https://github.com/microsoft/win32metadata/blob/273d404e3e851e31766a884c1bb9a14d3606570f/sources/GeneratorSdk/sdk/sdk.props#L40

Indeed, that works as well and I've switched to it now. I presume the compiler will complain on duplicate `using` statements, so that we won't forget about removing it when Mike's addition to `WinmdGenerator` ends up being released?

Reply:
That's great. Thanks for confirming.

Do you get a duplicate using error if you add Windows.Win32.Graphics.Dxgi to your own scraper header file? This using statement is already in the tooling that you're using.

Reply:
@mikebattista nothing, CI is green and no errors are seen when running the generator: https://github.com/MarijnS95/DirectX-Headers/commit/0a1d09598f9014e8c4cd1b280f9fc5485b678dee

Reply:
Ok. Then you likely won't see any conflicts when you update your tooling either. You can either keep it or just remember to remove it.